### DISCUSSION TITLE
Best way to test a Livewire Form object

### DISCUSSION THREAD

**Discussion by jeffchown** [livewire/livewire#8969] [1. Help]
Hello, all.

In my Livewire apps, I like to use one Form object for both create and edit purposes.

e.g. `App\Livewire\Client\Forms\ClientForm` which would contain, of course, properties, validation, etc. and methods for handling:
- `submit()`
- `store()`
- `update()`

I'd like to figure out a way to test the Form object on its own instead of having to test it through either e.g. my `App\Livewire\Client\CreateClient.php` or `App\Livewire\Client\EditClient.php` Livewire components.  That way, I know the Form object is working as expected no matter what Livewire component it is used in.

Is there a best Livewire practices/way to test a Form object in this way?

e.g. Is there a way to do something like you would for a component that would enable me to test my Form component independently?:

```
Livewire::test(ClientForm::class)
    ->set('code', 'TEST')
    -> ...;
```

**Comment by joshhanley** [comment]
@jeffchown hmm I think I just test them in the components, and just pick one component to add detailed tests to.

But you might be able to just test the FormObject as a unit test if you do like `$form = new FormObject($component, $propertyName);` But you will need a component object to pass into it, so not sure if it will work well or not. I think it only uses the component object for validation, so maybe you could just new up the base Livewire component and pass that in?

**Comment by jeffchown** [comment]
Thanks, @joshhanley - that is what I have done to-date (tested the component that uses the FormObject in question).

The more I use the same FormObjects in multiple components, the more I want to try to create a 'clean' test pattern to isolate the FormObject as much as possible when testing so it is not dependent or possibly affected by its containing component(s) during testing.

I am experimenting with exactly what you mention, by creating a 'minimalist' component and using that in the tests.

I wanted to confirm with you and @calebporzio that this is indeed the best way to perform such testing in case there was something I may have not known or overlooked.

e.g.:

```
<?php

use App\Livewire\Client\Forms\ClientForm;
use Livewire\Component;
use Livewire\Livewire;

beforeAll(function () {
    class ComponentWithClientForm extends Component
    {
        public ClientForm $form;

        public function submit(): void
        {
            $this->form->submit();
        }

        public function render(): string
        {
            return '<div></div>';
        }
    }
});

describe('Test ClientForm', function () {
    it('generates errors if required fields omitted', function () {
        // Arrange, Act & Assert
        Livewire::test(ComponentWithClientForm::class)
            ->call('submit')
            ->assertHasErrors(['form.code' => 'required'])
            ->assertHasErrors(['form.first_name' => 'required'])
            ->assertHasErrors(['form.last_name' => 'required']);
    });

    it('validates client code must be unique', function () {
        // Arrange, Act & Assert
        Livewire::test(ComponentWithClientForm::class)
            ...
    });

    it('creates a new client if the form ulid is empty', function () {
        // Arrange, Act & Assert
        Livewire::test(ComponentWithClientForm::class)
            ...
    });

    it('updates the client if the form ulid is *not* empty', function () {
        // Arrange, Act & Assert
        Livewire::test(ComponentWithClientForm::class)
            ...
    });
});
```

**Reply by lrljoe** [reply]
So I'd probably go a step further if you just want to test your Form objects and create a re-useable Component that you can set which Form you're testing.  Let me know if this needs clarification!

**Reply by jeffchown** [reply]
@lrljoe If I keep to something along the lines of my above pattern, a reusable component would definitely be a good next step. (and would eliminate the pesky error I get using the code above when running `composer dump-autoload` :)

**Reply by jeffchown** [reply]
@lrljoe Further to your suggestion, this is my current iteration of the pattern:

**TestCase.php**
``` 
<?php

namespace Tests;

use Illuminate\Foundation\Testing\TestCase as BaseTestCase;
use Livewire\Component;
use Livewire\Features\SupportFormObjects\Form;

abstract class TestCase extends BaseTestCase
{
    use CreatesApplication;

    protected $seed = true;
}

class ComponentWithForm extends Component
{
    public string $formClass = '';

    public Form $form;

    public function mount($formClass): void
    {
        $this->form = new $formClass($this, 'form');
    }

    public function submit(): void
    {
        if (method_exists($this->form, 'submit')) {
            $this->form->submit();
        }
    }

    public function render(): string
    {
        return '<div></div>';
    }
}
```

an example test for a `ClientForm` - **ClientFormTest.php**
```
<?php

use App\Livewire\Client\Forms\ClientForm;
use Livewire\Livewire;
use Tests\ComponentWithForm;

describe('Test ClientForm', function () {
    it('generates errors if required fields omitted', function () {
        // Arrange

        // Act & Assert
        Livewire::test(ComponentWithForm::class, [
            'formClass' => ClientForm::class,
        ])
            ->call('submit')
            ->assertHasErrors(['form.code' => 'required'])
            ->assertHasErrors(['form.first_name' => 'required'])
            ->assertHasErrors(['form.last_name' => 'required']);
    });
});
```

I'm liking this pattern so far.  Feels 'clean'.  Going to use it in a current project and see how it goes.

**Reply by lrljoe** [reply]
Yep, I go a step further and get the properties and validation rules for each form, and pass them to relevant handler methods that execute the various common tests.

Then I match that up against the fillable attributes on the Model to ensure they all have validation rules.

It works for the most part, with just a few specific extra methods for more complex validations/forms

**Reply by jeffchown** [reply]
@lrljoe Sounds good and thorough.  Have any examples you can share?

**Reply by lrljoe** [reply]
I'm phpunit rather than pest 

But effectively I just call the getRules() method on the form object to retrieve the rules.

Then for each rule, I have a method

e.g. testing 
min:X

tests
X-1
X
X+1

Then it's just a case of calling the correct method for every property on the form that has a "min" defined (with X changing per property)

Likewise for min/max it tests the same, but in combination, required, etc

I'd have to dig out the code as it's been a year since I even needed to look at it.
